bin/data_frame_analyzer/Main.cc

```diff
@@ -33,13 +33,33 @@
 #include <cstdlib>
 #include <fstream>
 #include <memory>
+#include <mutex>
 #include <string>
+#include <vector>
 
 namespace {
-std::pair<std::string, bool> readFileToString(const std::string& fileName) {
+auto makeErrorHandler(std::vector<std::string>& userErrors, std::mutex& userErrorsMutex) {
+    return [&userErrors, &userErrorsMutex](const std::string& error) {
+        try {
+            std::lock_guard<std::mutex> lock{userErrorsMutex};
+            userErrors.push_back(error);
+        } catch (std::exception& e) {
+            // We want this to be nothrow. Copying strings can fail for example with
+            // with std::bad_alloc so we catch and swallow any exceptions here.
```

Comment:
But logging an error could also fail with `std::bad_alloc` (and almost certainly will if copying the exact same string just failed due to lack of memory), so it won't be possible to log an error here if the intention is to make this lambda a nothrow.

I don't think eating `bad_alloc` exceptions is a good idea anyway.  We decided a long time ago to just let `bad_alloc` exceptions propagate to the top level of the stack and crash the program.  I have tried in the past to write handlers for out of memory and they never work as well in practice as the theory suggests.

If you leave `bad_alloc` unhandled then the C runtime eventually intervenes and prints an error to `stderr` in a way that's more reliable than higher level code.  The we have code in our `CppLogMessageHandler` Java class that picks up plain text messages from the `stderr` and logs them as fatal errors.

So I think the exception handler should just be removed here and we let any exception crash the process.

Reply:
Yes. My rationale was don't increase the chance of the process dying (although perhaps the comment is misleading). If we can't log without dying then fine go ahead and die, but don't add a possibly throwing operation in this context without at least trying to catch. However, this is all moot depending on how we want to proceed on communicating failure back to the calling process.